internal/controller/organization_controller.go

```
@@ -0,0 +1,153 @@
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: namespaceName,
+			Labels: map[string]string{
+				"giantswarm.io/organization": organization.Name,
+				"giantswarm.io/managed-by":   "organization-operator",
+			},
+		},
+	}
+
+	if err := ctrl.SetControllerReference(organization, namespace, r.Scheme); err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to set controller reference on Namespace: %w", err)
+	}
+
+	operationResult, err := ctrl.CreateOrUpdate(ctx, r.Client, namespace, func() error {
+		namespace.Labels = map[string]string{
+			"giantswarm.io/organization": organization.Name,
+			"giantswarm.io/managed-by":   "organization-operator",
+		}
+		return nil
+	})
+
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create or update Namespace: %w", err)
+	}
+
+	logger.Info("Namespace reconciled", "result", operationResult)
+
+	// Update Organization status
+	if organization.Status.Namespace != namespaceName {
+		patch := client.MergeFrom(organization.DeepCopy())
+		organization.Status.Namespace = namespaceName
+		if err := r.Status().Patch(ctx, organization, patch); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update Organization status: %w", err)
+		}
+	}
+
+	r.updateOrganizationCount(ctx)
+	return ctrl.Result{}, nil
+}
+
+func (r *OrganizationReconciler) reconcileDelete(ctx context.Context, organization *securityv1alpha1.Organization) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	originalOrg := organization.DeepCopy()
+	controllerutil.RemoveFinalizer(organization, "organization.giantswarm.io/finalizer")
+	patch := client.MergeFrom(originalOrg)
+	if err := r.Patch(ctx, organization, patch); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to remove finalizer: %w", err)
+	}
+	log.Info("Finalizer removed from Organization")
+
+	r.updateOrganizationCount(ctx)
+
+	return ctrl.Result{}, nil
+}
+func (r *OrganizationReconciler) updateOrganizationCount(ctx context.Context) {
```

let's return an error here and use normal error handling when calling this function

---

fixed